### DISCUSSION TITLE
Breaking change on nix

### DISCUSSION THREAD

**Discussion by noahfraiture** [hyprwm/Hyprland#8422] [Q&A] [answered]
Hello, 
A new breaking change block me from building my config on nixos. I get this error : 
```
error:
       … while calling the 'head' builtin
         at /nix/store/m68ikm8045fj7ys7qvgr608z9l70hh1k-source/lib/attrsets.nix:1575:11:
         1574|         || pred here (elemAt values 1) (head values) then
         1575|           head values
             |           ^
         1576|         else

       … while evaluating the attribute 'value'
         at /nix/store/m68ikm8045fj7ys7qvgr608z9l70hh1k-source/lib/modules.nix:816:9:
          815|     in warnDeprecation opt //
          816|       { value = addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          817|         inherit (res.defsFinal') highestPrio;

       … while evaluating the option `system.build.toplevel':

       … while evaluating definitions from `/nix/store/m68ikm8045fj7ys7qvgr608z9l70hh1k-source/nixos/modules/system/activation/top-level.nix':

       … while evaluating the option `xdg.portal.extraPortals':

       … while evaluating definitions from `/nix/store/m68ikm8045fj7ys7qvgr608z9l70hh1k-source/nixos/modules/programs/wayland/hyprland.nix':

       … while evaluating the option `programs.hyprland.portalPackage':

       … while evaluating definitions from `/nix/store/5hrs5cylad1xv137f6qjkqq9dq0zgqap-source/hosts/inspiron/hyprland.nix':

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)
   
```
which is not very explicit but I tried to remove the section mentionned to get this config :
```
{
  pkgs,
  inputs,
  ...
}:
{
  config = {
    nix.settings = {
      substituters = [ "https://hyprland.cachix.org" ];
      trusted-public-keys = [
        "hyprland.cachix.org-1:a7pgxzMz7+chwVL3/pzj6jIBMioiJM7ypFP8PwtkuGc="
      ];
    };

    programs.hyprland = {
      enable = true;
      package = inputs.hyprland.packages.${pkgs.system}.hyprland;
      xwayland.enable = true;
      # portalPackage =
      #   inputs.hyprland.packages.${pkgs.stdenv.hostPlatform.system}.xdg-desktop-portal-hyprland;
    };

    environment.sessionVariables = {
      NIXOS_OZONE_WL = "1";
      LIBVA_DRIVER_NAME = "nvidia";
      __GLX_VENDOR_LIBRARY_NAME = "nvidia";
    };

    # xdg.portal = {
    #   enable = true;
    #   wlr.enable = true;
    #   extraPortals = with pkgs; [
    #     xdg-desktop-portal-gtk
    #   ];
    # };

    services = {
      xserver = {
        enable = true;
        # displayManager.sessionCommands = ''
        #   ${pkgs.xorg.xrdb}/bin/xrdb -merge <<EOF
        #     Xft.dpi: 110
        #   EOF
        # '';
      };
    };
  };
}
```
 but again it doesn't work and I get this issue : 
```
warning: Git tree '/etc/nixos' is dirty
building the system configuration...
warning: Git tree '/etc/nixos' is dirty
error:
       … while calling the 'head' builtin
         at /nix/store/m68ikm8045fj7ys7qvgr608z9l70hh1k-source/lib/attrsets.nix:1575:11:
         1574|         || pred here (elemAt values 1) (head values) then
         1575|           head values
             |           ^
         1576|         else

       … while evaluating the attribute 'value'
         at /nix/store/m68ikm8045fj7ys7qvgr608z9l70hh1k-source/lib/modules.nix:816:9:
          815|     in warnDeprecation opt //
          816|       { value = addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          817|         inherit (res.defsFinal') highestPrio;

       … while evaluating the option `system.build.toplevel':

       … while evaluating definitions from `/nix/store/m68ikm8045fj7ys7qvgr608z9l70hh1k-source/nixos/modules/system/activation/top-level.nix':

       … while evaluating the option `warnings':

       … while evaluating definitions from `/nix/store/m68ikm8045fj7ys7qvgr608z9l70hh1k-source/nixos/modules/system/boot/systemd.nix':

       … while evaluating the option `systemd.services.home-manager-noah.serviceConfig':

       … while evaluating definitions from `/nix/store/m68ikm8045fj7ys7qvgr608z9l70hh1k-source/flake.nix':

       … while evaluating the option `home-manager.users.noah.home.activation.checkFilesChanged.data':

       … while evaluating definitions from `/nix/store/pij5lycb56qqa5r0ks43az07f2xfqr2n-source/modules/files.nix':

       … while evaluating the option `home-manager.users.noah.home.file."/home/noah/.config/hypr/hyprland.conf".onChange':

       … while evaluating definitions from `/nix/store/pij5lycb56qqa5r0ks43az07f2xfqr2n-source/modules/misc/xdg.nix':

       … while evaluating the option `home-manager.users.noah.xdg.configFile."hypr/hyprland.conf".onChange':

       … while evaluating definitions from `/nix/store/pij5lycb56qqa5r0ks43az07f2xfqr2n-source/modules/services/window-managers/hyprland.nix':

       … while evaluating the option `home-manager.users.noah.wayland.windowManager.hyprland.package':

       … while evaluating definitions from `/nix/store/w7lkb8sxbrxahnbw1xphl6cdma50zh7v-source/modules/home-manager/hyprland.nix':

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: path '/nix/store/fgxkr9w7f607kx637h5p5n1i0f51fj6g-source/flake.nix' does not exist
       
```

Does someone know what is the issue and what changed in hyprland ?

**Comment by fufexan** [comment]
Portals should be enabled and shouldn't cause the issue you're having.

What are your hyprland and nixpkgs inputs? Do you override anything? (`.follows`) Are your inputs up to date?

May also be useful to check which flake the last line belongs to. You can simply `ls /nix/store/fgxkr9w7f607kx637h5p5n1i0f51fj6g-source/`.

**Reply by noahfraiture** [reply]
My inputs are the following : 
```
    hyprland.url = "github:hyprwm/Hyprland";

    hyprspace = {
      url = "github:KZDKM/Hyprspace";
      inputs.hyprland.follows = "hyprland";
    };
```

Nothing has changed since my last build and I tried to checkout an old commit and it doesn't work either :(

The last flake is my global config, it gives the files inside my /etc/nixos

**Reply by fufexan** [reply]
A `--show-trace` might be useful in that case.

**Reply by noahfraiture** [reply]
Ok, I pasted it here because it's too long. To be honest I don't understand a lot from it https://pastebin.com/9TviZvLw

**Comment by noahfraiture** [comment]
The only solution I found to remove this error is to comment out `xdg-portal` `programs.hyprland.package` and `programs.hyprland.portalPackage` but then I get another error 
```
error:
       … while setting up the build environment

       error: getting attributes of path '/nix/store/26vlqxwzk1949vg41f7g56hqxa6ggq8r-make-shell-wrapper-hook': No such file or directory
``` when I deactivate all my home packages (containing hypridle and hyprlock) or 
```
error:
       … while calling the 'head' builtin
         at /nix/store/m68ikm8045fj7ys7qvgr608z9l70hh1k-source/lib/attrsets.nix:1575:11:
         1574|         || pred here (elemAt values 1) (head values) then
         1575|           head values
             |           ^
         1576|         else

       … while evaluating the attribute 'value'
         at /nix/store/m68ikm8045fj7ys7qvgr608z9l70hh1k-source/lib/modules.nix:816:9:
          815|     in warnDeprecation opt //
          816|       { value = addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          817|         inherit (res.defsFinal') highestPrio;

       … while evaluating the option `system.build.toplevel':

       … while evaluating definitions from `/nix/store/m68ikm8045fj7ys7qvgr608z9l70hh1k-source/nixos/modules/system/activation/top-level.nix':

       … while evaluating the option `warnings':

       … while evaluating definitions from `/nix/store/m68ikm8045fj7ys7qvgr608z9l70hh1k-source/nixos/modules/system/boot/systemd.nix':

       … while evaluating the option `systemd.services.home-manager-noah.serviceConfig':

       … while evaluating definitions from `/nix/store/m68ikm8045fj7ys7qvgr608z9l70hh1k-source/flake.nix':

       … while evaluating the option `home-manager.users.noah.home.activation.checkFilesChanged.data':

       … while evaluating definitions from `/nix/store/pij5lycb56qqa5r0ks43az07f2xfqr2n-source/modules/files.nix':

       … while evaluating the option `home-manager.users.noah.home.file."/home/noah/.config/hypr/hyprland.conf".onChange':

       … while evaluating definitions from `/nix/store/pij5lycb56qqa5r0ks43az07f2xfqr2n-source/modules/misc/xdg.nix':

       … while evaluating the option `home-manager.users.noah.xdg.configFile."hypr/hyprland.conf".onChange':

       … while evaluating definitions from `/nix/store/pij5lycb56qqa5r0ks43az07f2xfqr2n-source/modules/services/window-managers/hyprland.nix':

       … while evaluating the option `home-manager.users.noah.wayland.windowManager.hyprland.package':

       … while evaluating definitions from `/nix/store/8gfb24dr5rris70dghajwmrmsp92bj8s-source/modules/home-manager/hyprland.nix':
``` with all my home packages.
I'm a little bit lost since I had none of these issue before

**Reply by fufexan** [reply]
Do you have your config posted anywhere? Would be easier for me to check.

**Reply by noahfraiture** [reply]
> Do you have your config posted anywhere? Would be easier for me to check.

Yeah I turned it public https://github.com/noahfraiture/nix

**Reply by noahfraiture** [reply]
@fufexan Hey, did you see anything weird ?

**Reply by fufexan** [reply, marked as answer]
Hyprland is not the issue at all. It's a flake input that's not actually a flake, but you're trying to use it as such, and that's why Nix fails evaluation.

Applying this diff made both hosts build fine.
```diff
diff --git a/flake.nix b/flake.nix
index 54da3e3..01487be 100644
--- a/flake.nix
+++ b/flake.nix
@@ -26,7 +26,7 @@
 
     nix-flatpak.url = "github:gmodena/nix-flatpak/?ref=v0.4.1";
 
-    posting.url = "github:darrenburns/posting";
+    # posting.url = "github:darrenburns/posting";
 
     spicetify-nix = {
       url = "github:Gerg-L/spicetify-nix";
diff --git a/modules/home-manager/packages.nix b/modules/home-manager/packages.nix
index 630c402..0feacf0 100644
--- a/modules/home-manager/packages.nix
+++ b/modules/home-manager/packages.nix
@@ -40,12 +40,12 @@ let
     rainfrog # db management tool
     ripgrep # Search for substring
     yazi # file explorer
-    (python312.withPackages (
-      ps: with ps; [
-        posting
-      ]
-    ))
-    inputs.posting.packages.${pkgs.system}.default
+    # (python312.withPackages (
+    #   ps: with ps; [
+    #     posting
+    #   ]
+    # ))
+    # inputs.posting.packages.${pkgs.system}.default
 
   ];
 
```

**Reply by noahfraiture** [reply]
Sadly I still have the problem without that and I had it before trying to add posting. But if you say that it works on your side that's great ! I will make a full clear and rebuild from zero on a clean OS. Thank you for your help really !

**Reply by fufexan** [reply]
Ah, unfortunate situation. Best of luck!
